docs/csharp/advanced-topics/performance/index.md

```diff
@@ -0,0 +1,94 @@
+
+You can also use *ref assignment* to have a variable refer to another variable. A typical assignment copies the *value* of the right hand side to the variable on the left hand side of the assignment. A *ref assignment* copies the memory location of the variable on the right hand side to the variable on the left hand side. The `ref` now refers to the original variable:
+
+:::code language="csharp" source="./snippets/ref-safety/Program.cs" id="RefAssignment":::
+
+When you *assign* a variable, you change its *value*. When you *ref assign* a variable, you change what it refers to.
+
+You can work directly with the storage for values using `ref` variables, pass by reference, and ref assignment. Scope rules enforced by the compiler ensure safety when working directly with storage.
+
+## Ref safe to escape scope
+
+C# includes rules for `ref` expressions to ensure that a `ref` expression can't be accessed where the storage it refers to is no longer valid. Consider the following example:
+
+```csharp
+public ref int CantEscape()
+{
+    int index = 42;
+    return ref index; // Error: error's ref safe to escape scope is the body of CantEscape
+}
+```
+
+The compiler reports an error because you can't return a reference to a local variable from a method. The caller can't access the storage being referred to. The *ref safe to escape scope* defines the scope in which a `ref` expression is safe to access or modify. For non-ref locals, the *ref safe to escape* scope is the body of the method. In other words, a non-ref local can't be returned as a `ref` expression. For a variable to have a *ref safe to escape scope* of the entire method, it must be accessible and assignable outside the method. The following snippet shows two examples. A member field can be accessed from the scope calling a method, so a class or struct field's *ref safe to escape scope* is the entire method. The *ref safe to escape scope* for a parameter with the `ref`, `in`, or `out` modifiers is also the entire method. Both can be `ref` returned from a member method:
+
+:::code language="csharp" source="./snippets/ref-safety/EscapeScopes.cs" id="RefSafeToEscapeScopes":::
+
+> [!NOTE]
+> When the `in` modifier is applied to a parameter, that parameter can be returned by `ref readonly`, not `ref`.
+
+The compiler ensures that a reference can't escape its *ref safe to escape scope*. You can use `ref` parameters, `ref return` and `ref` local variables safely because the compiler detects if you've accidentally written code where a `ref` expression could be accessed when its storage isn't valid.
+
+## Safe to escape scope and ref structs
+
+`ref struct` types require more rules to ensure they can be used safely. A `ref struct` type may include `ref` fields. That requires the introduction of a *safe to escape scope*. For most types, the *safe to escape scope* is the entire method. In other words, a value that's not a `ref struct` can always be returned from a method.
```

Comment:
>the *safe to escape scope* is the entire method [](http://example.com/codeflow?start=195&length=47)

For types other than ref structs, the safe to escape scope is the _calling method_. (A value with safe to escape scope of _calling method_ is safe to return from the method; a value with safe to escape scope of current method is not.)